src/outputs/parthenon_hdf5.cpp

```diff
@@ -192,436 +330,652 @@ void PHDF5Output::genXDMF(std::string hdfFile, Mesh *pm, SimTime *tm) {
-      } else {
-        vLocalSpace = vLocalSpaceNew =
-            H5S::FromHIDCheck(H5Screate_simple(5, local_count, NULL));
-        vGlobalSpace = vGlobalSpaceNew =
-            H5S::FromHIDCheck(H5Screate_simple(5, global_count, NULL));
-      }
+  }
+
+  // -------------------------------------------------------------------------------- //
+  //   WRITING VARIABLES DATA                                                         //
+  // -------------------------------------------------------------------------------- //
+
+  // first we need to get list of variables, because sparse variables are only
+  // expanded on some blocks, we need to look at the list of variables on each block
+  // combine these into a global list of variables
+
+  auto get_MeshBlockDataIterator = [=](const std::shared_ptr<MeshBlock> pmb) {
+    if (restart_) {
+      return MeshBlockDataIterator<Real>(
+          pmb->meshblock_data.Get(),
+          {parthenon::Metadata::Independent, parthenon::Metadata::Restart}, true);
+    } else {
+      return MeshBlockDataIterator<Real>(pmb->meshblock_data.Get(),
+                                         output_params.variables);
+    }
+  };
+
+  // local list of unique vars
+  std::set<VarInfo> all_unique_vars;
+  for (auto &pmb : pm->block_list) {
+    auto ci = get_MeshBlockDataIterator(pmb);
+    for (auto &v : ci.vars) {
+      VarInfo vinfo(v);
+      all_unique_vars.insert(vinfo);
+    }
+  }
+
+#ifdef MPI_PARALLEL
+  {
+    // we need to do a global allgather to get the global list of unique variables to
+    // be written to the HDF5 file
+
+    // the label buffer contains all labels of the unique variables on this rank
+    // separated by \t, e.g.: "label0\tlabel1\tlabel2\t"
+    std::vector<std::string> compact_labels;
+    std::vector<int> code_buffer;
+
+    for (const auto &vi : all_unique_vars) {
+      compact_labels.push_back(vi.get_compact_label());
+      code_buffer.push_back(vi.get_info_code());
+    }
+
+    std::string label_buffer = string_utils::PackStrings(compact_labels, '\n');
+
+    // first we need to communicate the lengths of the label_buffer and vlen_buffer to
+    // all ranks, 2 ints per rank: first int: label_buffer length, second int:
+    // vlen_buffer length
+    std::vector<int> buffer_lengths(2 * Globals::nranks, 0);
+    buffer_lengths[Globals::my_rank * 2 + 0] = static_cast<int>(label_buffer.size());
+    buffer_lengths[Globals::my_rank * 2 + 1] = static_cast<int>(code_buffer.size());
+
+    PARTHENON_MPI_CHECK(MPI_Allgather(MPI_IN_PLACE, 2, MPI_INT, buffer_lengths.data(), 2,
+                                      MPI_INT, MPI_COMM_WORLD));
+
+    // now do an Allgatherv combining label_buffer and vlen_buffer from all ranks
+    std::vector<int> label_lengths(Globals::nranks, 0);
+    std::vector<int> label_offsets(Globals::nranks, 0);
+    std::vector<int> code_lengths(Globals::nranks, 0);
+    std::vector<int> code_offsets(Globals::nranks, 0);
```

Comment:
You are assigning 0's then overwriting them, I would use reserve here with push back instead.

Reply:
Since the size is fixed and known ahead of time, I think this is slightly more straightforward and readable than using `reserve` and `push_back`